src/GraphQL/Validation/Rules/NoUndefinedVariables.cs

```diff
@@ -20,28 +20,29 @@ public class NoUndefinedVariables : IValidationRule
 
         /// <inheritdoc/>
         /// <exception cref="NoUndefinedVariablesError"/>
-        public Task<INodeVisitor> ValidateAsync(ValidationContext context)
-        {
-            var variableNameDefined = new Dictionary<string, bool>();
+        public Task<INodeVisitor> ValidateAsync(ValidationContext context) => _nodeVisitor;
 
-            return new EnterLeaveListener(_ =>
+        private static readonly Task<INodeVisitor> _nodeVisitor = new NodeVisitors(
+            new MatchingNodeVisitor<VariableDefinition>((varDef, context) =>
             {
-                _.Match<VariableDefinition>(varDef => variableNameDefined[varDef.Name] = true);
+                var varNameDef = context.TypeInfo.NoUndefinedVariables_VariableNameDefined ??= new HashSet<string>();
+                varNameDef.Add(varDef.Name);
+            }),
 
-                _.Match<Operation>(
-                    enter: op => variableNameDefined = new Dictionary<string, bool>(),
-                    leave: op =>
+            new MatchingNodeVisitor<Operation>(
+                enter: (op, context) => context.TypeInfo.NoUndefinedVariables_VariableNameDefined?.Clear(),
```

Comment:
Dictionaries are cleared instead of allocating new instances when necessary.  If null it just stays null.